File: backend/app/models/financial.py
from pydantic import BaseModel, validator
from decimal import Decimal
from datetime import datetime
from typing import Optional, List
from enum import Enum

class TransactionType(str, Enum):
    DEBIT = "debit"
    CREDIT = "credit"

class TransactionCategory(str, Enum):
    GROCERIES = "groceries"
    DINE_OUT = "dine_out"
    ENTERTAINMENT = "entertainment"
    BILLS = "bills"
    TRANSPORT = "transport"
    SHOPPING = "shopping"
    HEALTH = "health"
    FITNESS = "fitness"
    SAVINGS = "savings"
    INCOME = "income"
    OTHER = "other"

class TransactionCreate(BaseModel):
    amount: Decimal
    category: TransactionCategory
    description: str
    merchant_name: Optional[str] = None
    location_city: Optional[str] = "Unknown"
    location_country: str = "US"

    @validator('amount')
    def validate_amount(cls, v):
        if v == 0:
            raise ValueError('Transaction amount cannot be zero')
        return v

class TransactionResponse(BaseModel):
    id: str
    transaction_type: TransactionType
    amount: Decimal
    category: str
    description: str
    merchant_name: Optional[str]
    location_city: Optional[str]
    location_country: str
    balance_after: Decimal
    transaction_date: datetime

    class Config:
        from_attributes = True

class WalletResponse(BaseModel):
    id: str
    balance: Decimal
    currency: str
    updated_at: datetime

    class Config:
        from_attributes = True

class SpendingAnalytics(BaseModel):
    category: str
    total_amount: Decimal
    transaction_count: int
    percentage: float

class FinancialHealthSummary(BaseModel):
    current_balance: Decimal
    weekly_spending: Decimal
    weekly_income: Decimal
    transaction_count: int
    savings_rate: float
    daily_spending_avg: Decimal
    top_spending_categories: List[SpendingAnalytics]

class TransactionFilter(BaseModel):
    limit: Optional[int] = 50
    days: Optional[int] = None
    category: Optional[str] = None
    transaction_type: Optional[TransactionType] = None
    min_amount: Optional[Decimal] = None
    max_amount: Optional[Decimal] = None

    @validator('limit')
    def validate_limit(cls, v):
        if v and (v < 1 or v > 100):
            raise ValueError('Limit must be between 1 and 100')
        return v

File: frontend/src/types/financial.ts
// Financial-related types matching backend models

export type TransactionType = "debit" | "credit";

export type TransactionCategory =
  | "groceries"
  | "dine_out"
  | "entertainment"
  | "bills"
  | "transport"
  | "shopping"
  | "health"
  | "fitness"
  | "savings"
  | "income"
  | "other";

export interface TransactionCreate {
  amount: number;
  category: TransactionCategory;
  description: string;
  merchant_name?: string;
  location_city?: string;
  location_country?: string;
}

export interface Transaction {
  id: string;
  transaction_type: TransactionType;
  amount: number;
  category: string;
  description: string;
  merchant_name?: string;
  location_city?: string;
  location_country: string;
  balance_after: number;
  transaction_date: Date;
}

export interface Wallet {
  id: string;
  balance: number;
  currency: string;
  updated_at: Date;
}

export interface SpendingAnalytics {
  category: string;
  total_amount: number;
  transaction_count: number;
  percentage: number;
}

export interface FinancialHealthSummary {
  current_balance: number;
  weekly_spending: number;
  weekly_income: number;
  transaction_count: number;
  savings_rate: number;
  daily_spending_avg: number;
  top_spending_categories: SpendingAnalytics[];
}

export interface TransactionFilter {
  limit?: number;
  days?: number;
  category?: string;
  transaction_type?: TransactionType;
  min_amount?: number;
  max_amount?: number;
}

// Frontend-specific financial types
export interface TransactionFormData {
  amount: string; // String for form input
  category: TransactionCategory;
  description: string;
  merchant_name?: string;
  location_city?: string;
  location_country?: string;
}

export interface TransactionListState {
  transactions: Transaction[];
  isLoading: boolean;
  error: string | null;
  hasMore: boolean;
  page: number;
}

export interface WalletState {
  wallet: Wallet | null;
  isLoading: boolean;
  error: string | null;
  lastUpdated: Date | null;
}

export interface FinancialFilters {
  dateRange: {
    start: Date | null;
    end: Date | null;
  };
  categories: TransactionCategory[];
  transactionTypes: TransactionType[];
  amountRange: {
    min: number | null;
    max: number | null;
  };
  searchTerm: string;
}

export interface SpendingTrend {
  date: Date;
  amount: number;
  category?: string;
}

export interface BudgetGoal {
  id: string;
  category: TransactionCategory;
  monthly_limit: number;
  current_spent: number;
  percentage_used: number;
  is_exceeded: boolean;
}

export interface FinancialInsight {
  type: "warning" | "success" | "info";
  title: string;
  message: string;
  category?: TransactionCategory;
  amount?: number;
  suggestion?: string;
}

export interface MonthlyReport {
  month: string;
  year: number;
  total_income: number;
  total_expenses: number;
  net_savings: number;
  category_breakdown: SpendingAnalytics[];
  insights: FinancialInsight[];
}

// Analytics-specific types
export interface BalanceHistoryPoint {
  date: string;
  balance: number;
  transaction_type: TransactionType;
  amount: number;
}

export interface MonthlySummary {
  month: number;
  year: number;
  total_income: number;
  total_spending: number;
  net_change: number;
  transaction_count: number;
  average_transaction: number;
}

export interface AnalyticsTimeRange {
  label: string;
  days: number;
  value: string;
}

export interface FinancialMetric {
  label: string;
  value: string | number;
  change?: number;
  changeType?: "increase" | "decrease" | "neutral";
  format: "currency" | "percentage" | "number";
  status?: "good" | "warning" | "danger";
}

export interface SpendingTrendData {
  date: string;
  spending: number;
  income: number;
  balance: number;
}

export interface CategorySpendingData {
  category: string;
  amount: number;
  percentage: number;
  color: string;
  transactions: number;
}
